cmake/external/dml.cmake

```diff
@@ -20,19 +20,17 @@ if (NOT onnxruntime_USE_CUSTOM_DIRECTML)
   set(NUGET_CONFIG ${PROJECT_SOURCE_DIR}/../NuGet.config)
   set(PACKAGES_CONFIG ${PROJECT_SOURCE_DIR}/../packages.config)
   set(PACKAGES_DIR ${CMAKE_CURRENT_BINARY_DIR}/packages)
+  set(DML_PACKAGE_DIR ${PACKAGES_DIR}/DirectML.0.0.1)
 
   # Restore nuget packages, which will pull down the DirectML redist package
   add_custom_command(
-    OUTPUT restore_packages.stamp
+    OUTPUT ${DML_PACKAGE_DIR}/bin/x64/DirectML.lib ${DML_PACKAGE_DIR}/bin/x86/DirectML.lib
```

Comment:
Ninja failed with ninja: `error: 'packages/DirectML.0.0.1/bin/x64/DirectML.lib', needed by 'onnxruntime.lib', missing and no known rule to make it`, given that it didn't know how to get DirectML.lib which was a dep of ORT.

With msbuild it just happened to not check if it could satisfy all dependencies first and happened to run the RESTORE_PACKAGES target before building ORT.

Reply:
Here I'm declaring that this command will output x64/DirectML.lib and x86/DirectML.lib ; if any target depends on either of those, the package will be downloaded using Nuget. There's a single nuget package with both, no option to pull them separatly.

ARM is not supported, I was talking to Sheil about it last week. DML distributes a redist package with x86 and x64 DML, but this public redist package doesn't have ARM (seems like they didn't have a need for it and didn't want to build and maintain it if no customers are using it. We were not sure if that's the only reason, we can talk to the DML folks if we need an ARM redist package)

Reply:
It looks like the intention was to have onnxruntime depend on onnxruntime_EXTERNAL_DEPENDENCIES, which implicitly includes the DML libs. This change looks fine too, and it's more explicit.

Reply:
Yes, that's right, but it doesn't work with ninja. ORT links with DirectML.lib, which doesn't exist when doing a clean build because the nuget package isn't downloaded yet, and there is no rule on how to generate DirectML.lib, so it says `error: 'packages/DirectML.0.0.1/bin/x64/DirectML.lib', needed by 'onnxruntime.lib', missing and no known rule to make it`.

Depending on onnxruntime_EXTERNAL_DEPENDENCIES will eventually pull DirectML.lib, but there's no way the build system knows about it before pulling and unpacking the package. It seems like msbuild doesn't validate that it can actually resolve everything before starting the build, so it worked before.

Reply:
When I originally set this up, cmake actually had a similar problem. It barfed when trying to target_link_libraries the DirectML.targets file, because the nuget package hadn't been downloaded yet (and therefore the targets file didn't exist).

This was solved by adding an explicit step into build.py which invokes the RESTORE_PACKAGES step prior to running the "real" build. This ensures that when the build starts, the nuget package is already present.

Why doesn't that solution work in this case?

Reply:
Oh sure, that also works. I didn't realize RESTORE_PACKAGES was called in build.py.

I was running the ninja build "by hand" (calling cmake followed by cmake --build or configuring the cmake project in VS directly).

Is there a reason not to fix it the way I did?